This document demonstrates what I needed to do to instal the [COBRA Toolbox](http://opencobra.github.io/cobratoolbox/) and get it running on a fresh MATLAB installation.

Some technical details: I'm using a Mac running OS X Version 10.11.3 and MATLAB version 2014a. If you're in a different environment, you'll likely need to adjust your path and other details accordingly.

If you need help, the [COBRA Toolbox Google Group](https://groups.google.com/forum/#!forum/cobra-toolbox) is a good place to ask.

To start, I'll check that this is a fresh install (there's no COBRA Toolbox or glpk):

In [3]:
which initCobraToolbox;

'initCobraToolbox' not found.


In [4]:
which glpk;

'glpk' not found.


I'll begin by checking the local working directory and its contents:

In [21]:
cd

/Users/bheavner/Projects/cobra_jupyter


In [2]:
ls -l

total 40
-rw-r--r--  1 bheavner  staff  9494 Feb  9 14:10 Installing the COBRA Toolbox.ipynb
-rw-r--r--  1 bheavner  staff  1078 Feb  9 12:51 LICENSE
-rw-r--r--  1 bheavner  staff  1536 Feb  9 13:31 hello_world.ipynb


Now I'll download the COBRA Toolbox. For convienece, I'll do this in the command lime instead of in this notebook. The -L flag on curl is needed because the download redirects. If you prefer, you can just going to the dowhload page and clicking the link. The command I'll use is:

`curl -O -J -L https://github.com/opencobra/cobratoolbox/tarball/master`

Did it work?

In [1]:
ls -l

total 329936
-rw-r--r--  1 bheavner  staff       5242 Feb  9 14:19 Installing the COBRA Toolbox.ipynb
-rw-r--r--  1 bheavner  staff       1078 Feb  9 12:51 LICENSE
-rw-r--r--  1 bheavner  staff       1536 Feb  9 13:31 hello_world.ipynb
-rw-r--r--  1 bheavner  staff  168910051 Feb  9 14:22 opencobra-cobratoolbox-4482ec0.tar.gz


Next, unzip the toolbox. Again, I'll do it at the command line instead of this notebook. I'll use

`tar -xvf opencobra-cobratoolbox-4482ec0.tar.gz`

In [2]:
ls -l

total 329936
-rw-r--r--    1 bheavner  staff       5242 Feb  9 14:19 Installing the COBRA Toolbox.ipynb
-rw-r--r--    1 bheavner  staff       1078 Feb  9 12:51 LICENSE
-rw-r--r--    1 bheavner  staff       1536 Feb  9 13:31 hello_world.ipynb
drwxr-xr-x  103 bheavner  staff       3502 Feb  4 13:51 opencobra-cobratoolbox-4482ec0
-rw-r--r--    1 bheavner  staff  168910051 Feb  9 14:22 opencobra-cobratoolbox-4482ec0.tar.gz


In [3]:
cd opencobra-cobratoolbox-4482ec0

In [4]:
cd

/Users/bheavner/Projects/cobra_jupyter/opencobra-cobratoolbox-4482ec0


So that's it - we've got the COBRA Toolbox now. Test it.

In [5]:
initCobraToolbox

[Warning: Function isrow has the same name as a MATLAB builtin. We suggest you
rename the function to avoid a potential name conflict.] 
[> In path at 109
  In addpath at 86
  In addpath_recurse at 35
  In addpath_recurse at 43
  In initCobraToolbox at 30
  In pymat_eval at 31
  In matlabserver at 24] 
Define LP solver...
[Warning: LP solver Gurobi not useable: gurobi.m not in Matlab path] 
[> In changeCobraSolver at 160
  In initCobraToolbox at 48
  In pymat_eval at 31
  In matlabserver at 24] 
[Warning: LP solver Gurobi not useable: gurobi.m not in Matlab path] 
[> In changeCobraSolver at 155
  In initCobraToolbox at 48
  In pymat_eval at 31
  In matlabserver at 24] 
[Warning: LP solver Gurobi not useable: gurobi_mex not in Matlab path] 
[> In changeCobraSolver at 150
  In initCobraToolbox at 48
  In pymat_eval at 31
  In matlabserver at 24] 
[Warning: LP solver CPLEX through Tomlab not usable: tomRun.m not in Matlab path] 
[> In changeCobraSolver at 123
  In initC

I'm not sure why I'm getting that "Function isrow has the same name as a MATLAB builtin" warning.. it appears to come from:
>In path at 109
>  In addpath at 86
>  In addpath_recurse at 35
>  In addpath_recurse at 43
>  In initCobraToolbox at 30 

and that last line is a surprise! 

The COBRA Toolbox README says "To read / write SBML files you must install libSBML 5.5.0 from here: http://sourceforge.net/projects/sbml/files/libsbml/" - I don't recall doing that before on this machine, but translateSBML seems to work anyway? Ah well. I'll proceed!

Next, test the install (note: I don't think I've installed the sbml toolbox or a solver at this point, so expect the tests using glpk to work, but not others...)

In [ ]:
testAll

This seems to timeout on the "single reaction deletion analysis" test, so I jump from this notebook over to MATLAB and run it there. After they run, the output is this:

```
Tests passed: 
testBuildMPS
testDeletionStudy
testElementalBalance
testFBA
testFVA
testFastcc
testFluxConsistency
testGrowthExpMatch
testMaps
testModelManipulation
modelCitations_update
testRank
testRobustnessAnalysis
testgpSampler
testSubspaces
testTissueModel


Tests not passed: 
testC13Fitting
testFASTCORE
testGDLS
testMOMA
modelCitations
sbmlTestModelToMat
testModels
testModelsAlgorithmically
testNullspace
testOptKnock
testSBML
testSampleCbModel
testDifferentLPSolvers
testSolvers
testStoichiometricConsistency
testpFBA
IT IS NOT NECESSARY FOR THE COBRA TOOLBOX TO PASS ALL TESTS TO FUNCTION; HOWEVER, IT MUST PASS THE TESTS THAT ARE RELEVANT TO YOUR PARTICULAR PROBLEM!!!
Tests may not pass for several reasons.  Some of the most common issues:
1.  The correct solver is not installed.  Certain tests require LP, MILP, QP or NLP solvers.  See changeCobraSolvers.m for a complete list of supported solvers.
These tests will fail when running testAll unless one has the tomlab suite installed.  If all of the functions that you require for your use function then do not worry about them: testC13Fitting, testGDLS, testMOMA, testOptKnock, testSolvers
If a particular test fails, you can run that test individually for more information

ans =

16 of 32 tests completed successfully.
```

Cool! I'll want to use SBML, so I'll go back to install the SBML Toolbox from http://sourceforge.net/projects/sbml/files/libsbml/ - I'll grab the 5.12.0 stable release mac OS X binary and install it. Specifically, I want the MATLAB binding. The OS X version is at http://sourceforge.net/projects/sbml/files/libsbml/5.12.0/stable/Mac%20OS%20X/MATLAB%20interface/

The documentation says:
Installation instructions

>    Extract the archive somewhere on your local computer.
>
>    Start your copy of MATLAB and add the folder to your MATLAB path using addPath.
>
>    Verify that everything works. You can do this easily by changing into that folder within your running MATLAB environment, then issuing the command TranslateSBML('test.xml') in MATLAB. If everything went as planned, the output of that command will be a MATLAB structure representing the toy test model.
>
>    Save the changes to your MATLAB path for future use. To do this, you can use the savepath command in MATLAB.

So I'll go to my command line and change directory to /Users/bheavner/Projects/cobra_jupyter then download the software with 

`curl -O -J -L http://sourceforge.net/projects/sbml/files/libsbml/5.12.0/stable/Mac%20OS%20X/MATLAB%20interface/libSBML-5.12.0-macosx-matlab-binaries.tar.gz/download`

In [1]:
cd /Users/bheavner/Projects/cobra_jupyter

In [2]:
cd

/Users/bheavner/Projects/cobra_jupyter


In [3]:
ls -l

total 335352
-rw-r--r--    1 bheavner  staff      14880 Feb  9 15:33 Installing the COBRA Toolbox.ipynb
-rw-r--r--    1 bheavner  staff       1078 Feb  9 12:51 LICENSE
-rw-r--r--    1 bheavner  staff       1536 Feb  9 13:31 hello_world.ipynb
-rw-r--r--    1 bheavner  staff    2762569 Feb  9 15:34 libSBML-5.12.0-macosx-matlab-binaries.tar.gz
drwxr-xr-x  103 bheavner  staff       3502 Feb  4 13:51 opencobra-cobratoolbox-4482ec0
-rw-r--r--    1 bheavner  staff  168910051 Feb  9 14:22 opencobra-cobratoolbox-4482ec0.tar.gz


Unzip it with `tar -xvf libSBML-5.12.0-macosx-matlab-binaries.tar.gz` to get:

In [5]:
ls -l

total 335360
-rw-r--r--    1 bheavner  staff      16557 Feb  9 15:35 Installing the COBRA Toolbox.ipynb
-rw-r--r--    1 bheavner  staff       1078 Feb  9 12:51 LICENSE
-rw-r--r--    1 bheavner  staff       1536 Feb  9 13:31 hello_world.ipynb
-rw-r--r--    1 bheavner  staff    2762569 Feb  9 15:34 libSBML-5.12.0-macosx-matlab-binaries.tar.gz
drwxr-xr-x   16 bheavner  staff        544 Nov 10 07:08 libSBML-5.12.0-matlab
drwxr-xr-x  103 bheavner  staff       3502 Feb  4 13:51 opencobra-cobratoolbox-4482ec0
-rw-r--r--    1 bheavner  staff  168910051 Feb  9 14:22 opencobra-cobratoolbox-4482ec0.tar.gz


Add it to the MATLAB Path, and save the path:

In [6]:
addpath('/Users/bheavner/Projects/cobra_jupyter/libSBML-5.12.0-matlab')

test the install:

In [7]:
cd libSBML-5.12.0-matlab; TranslateSBML('test.xml')

ans = 

              typecode: 'SBML_MODEL'
                metaid: ''
                 notes: '<notes>
  <body xmlns="http://www.w3.org/1999/xht...'
            annotation: ''
            SBML_level: 2
          SBML_version: 1
                  name: ''
                    id: 'Branch'
    functionDefinition: [1x0 struct]
        unitDefinition: [1x0 struct]
           compartment: [1x1 struct]
               species: [1x4 struct]
             parameter: [1x0 struct]
                  rule: [1x0 struct]
              reaction: [1x3 struct]
                 event: [1x0 struct]
           time_symbol: ''
          delay_symbol: ''
            namespaces: [1x1 struct]


Save the changes to your MATLAB path for future use. 

In [10]:
savepath

Then test the SBML i/o functionality in the cobraToolbox

In [16]:
cd opencobra-cobratoolbox-4482ec0

In [19]:
testSBML

Undefined function 'Model_create' for input arguments of type 'double'.


Running this again in the MATLAB environment, I see that this error comes from the following:

>`>>> testSBML`
>Test for readCbModel succeeded
>Undefined function 'Model_create' for input arguments of type 'double'.
>
>Error in convertCobraToSBML (line 63)
>sbmlModel = Model_create(sbmlLevel, sbmlVersion);
>
>
>Error in writeCbModel (line 256)
>        sbmlModel = convertCobraToSBML(model,sbmlLevel,sbmlVersion,compSymbolList,compNameList);
>
>
>Error in testSBML (line 92)
>writeCbModel( testModel, 'sbml', 'test_model.sbml');

I mostly care about reading SBML files at the moment, not writing. So that error is okay for me!